src/tests/roc_audio/test_watchdog.cpp

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2017 Mikhail Baranov
+ * Copyright (c) 2017 Victor Gaydov
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <CppUTest/TestHarness.h>
+
+#include "roc_audio/watchdog.h"
+#include "roc_core/buffer_pool.h"
+#include "roc_core/heap_allocator.h"
+
+namespace roc {
+namespace audio {
+
+namespace {
+
+enum {
+    MaxBufSize = 5000,
+
+    SamplesPerPacket = 200,
+    NumCh = 2,
+
+    Timeout = 20
+};
+
+core::HeapAllocator allocator;
+core::BufferPool<sample_t> sample_buffer_pool(allocator, MaxBufSize, 1);
+
+class TestFrameReader : public IReader, public core::NonCopyable<> {
+public:
+    void read(Frame& frame) {
+        memset(frame.samples().data(), 0, frame.samples().size() * sizeof(sample_t));
+    }
+};
+
+TestFrameReader test_reader;
+
+} // namespace
+
+TEST_GROUP(watchdog) {
+    Frame new_frame(size_t sz) {
+        core::Slice<sample_t> samples =
+            new (sample_buffer_pool) core::Buffer<sample_t>(sample_buffer_pool);
+        samples.resize(sz * NumCh);
+        Frame frame(samples);
+        return frame;
+    }
+
+    void check_read(unsigned frame_flags, IReader& reader, bool is_read) {
+        Frame frame = new_frame(SamplesPerPacket);
+        frame.add_flags(frame_flags);
+
+        reader.read(frame);
+
+        if (is_read) {
+            for (size_t n = 0; n < frame.samples().size(); n++) {
+                DOUBLES_EQUAL(0.0f, frame.samples().data()[n], 0);
```

Comment:
This doesn't guarantee that Watchdog actually performed a read. Better fill the frame with non-zero values in TestFrameReader and then check them here. Also it's make sense to check for zero values when `is_read` is false.